staging/src/k8s.io/apiserver/pkg/endpoints/handlers/apply.go

```diff
@@ -19,162 +19,66 @@ package handlers
-	"k8s.io/apimachinery/pkg/util/json"
+	"k8s.io/apiserver/pkg/endpoints/handlers/apply"
 	"k8s.io/kube-openapi/pkg/util/proto"
+	"sigs.k8s.io/structured-merge-diff/fieldpath"
+	"sigs.k8s.io/structured-merge-diff/merge"
+	"sigs.k8s.io/structured-merge-diff/typed"
 )
 
 type applyPatcher struct {
 	*patcher
 
-	models proto.Models
+	models    proto.Models
+	converter apply.TypeConverter
+	updater   merge.Updater
 }
 
-// TODO(apelisse): workflowId needs to be passed as a query
-// param/header, and a better defaulting needs to be defined too.
-const workflowId = "default"
+const applyManager = "apply"
 
-func (p *applyPatcher) convertCurrentVersion(obj runtime.Object) (map[string]interface{}, error) {
-	vo, err := p.unsafeConvertor.ConvertToVersion(obj, p.hubGroupVersion)
+func (p *applyPatcher) applyPatchToCurrentObject(currentObject runtime.Object) (runtime.Object, error) {
+	managed, err := apply.DecodeObjectManagedFields(currentObject)
 	if err != nil {
 		return nil, err
 	}
-	return runtime.DefaultUnstructuredConverter.ToUnstructured(vo)
-}
-
-func (p *applyPatcher) extractLastIntent(obj runtime.Object, workflow string) (map[string]interface{}, error) {
-	accessor, err := meta.Accessor(obj)
-	if err != nil {
-		return nil, fmt.Errorf("couldn't get accessor: %v", err)
-	}
-	last := make(map[string]interface{})
-	// TODO: use the managedFields correctly
-	if _, ok := accessor.GetManagedFields()[workflow]; ok {
-		if err := json.Unmarshal([]byte(accessor.GetManagedFields()[workflow].APIVersion), &last); err != nil {
-			return nil, fmt.Errorf("couldn't unmarshal managedFields field: %v", err)
-		}
-	}
-	return last, nil
-}
-
-func (p *applyPatcher) getNewIntent() (map[string]interface{}, error) {
-	patch := make(map[string]interface{})
-	if err := yaml.Unmarshal(p.patchBytes, &patch); err != nil {
-		return nil, fmt.Errorf("couldn't unmarshal patch object: %v (patch: %v)", err, string(p.patchBytes))
-	}
-	return patch, nil
-}
 
-func (p *applyPatcher) convertResultToUnversioned(result apply.Result) (runtime.Object, error) {
-	voutput, err := p.creater.New(p.kind)
+	currentTyped, err := p.converter.ObjectToTyped(currentObject)
 	if err != nil {
-		return nil, fmt.Errorf("failed to create empty output object: %v", err)
+		return nil, fmt.Errorf("failed to create typed current object: %v", err)
 	}
 
-	err = runtime.DefaultUnstructuredConverter.FromUnstructured(result.MergedResult.(map[string]interface{}), voutput)
+	newTyped, err := p.converter.YAMLToTyped(p.patchBytes)
 	if err != nil {
-		return nil, fmt.Errorf("failed to convert merge result back: %v", err)
+		return nil, fmt.Errorf("failed to convert patch to typed object: %v", err)
 	}
-	p.defaulter.Default(voutput)
 
-	gvk := p.kind.GroupKind().WithVersion(runtime.APIVersionInternal)
-	uoutput, err := p.unsafeConvertor.ConvertToVersion(voutput, gvk.GroupVersion())
+	outputTyped, managed, err := p.updater.Apply(currentTyped, newTyped, managed, applyManager, *p.options.Force)
```

Comment:
You can't dereference `p.options.Force` if it's nil, you need to default to False if not set. (unless it's defaulter, but I really don't think it is)